### PR TITLE
#988 Separate decodeCborHexToBytes and remove textEnvelopeBytes in favor of plutusScriptVxFromEnvelope

### PR DESCRIPTION
**Pull request by uhbif19**
Not sure if that is what was meant in [the issue](https://github.com/Plutonomicon/cardano-transaction-lib/issues/988).

### Pre-review checklist

- [x] All code has been formatted using our config (`make format`)
- [x] Any new API features or modification of existing behavior is covered as defined in our [test plan](https://github.com/Plutonomicon/cardano-transaction-lib/blob/develop/doc/test-plan.md)
- [x] The changelog has been updated under the `## Unreleased` header, using the appropriate sub-headings (`### Added`, `### Removed`, `### Fixed`), and the links to the appropriate issues/PRs have been included

### CODE REVIEW

**Comment by uhbif19** [comment]
@klntsky 

Not sure I understand task correctly.

I used signature `String -> Either TextEnvelopeDecodeError PlutusScript` cuz using Aeson/Maybe leads to long boilerplate code like this:

```
alwaysFailsScript :: Contract () Validator
alwaysFailsScript = do
  let aesonOrLeft = parseJsonStringToAeson alwaysFails
  let aesonError = error "Cannot decode JSON for alwaysFailScript"
  let fromEnvelopeError = error "Cannot decode envelope for alwaysFailScript"
  aeson <- liftEither $ lmap (const aesonError) aesonOrLeft
  liftMaybe fromEnvelopeError $ Validator <$> plutusScriptV1FromEnvelope aeson
```

Also, I did this: https://github.com/Plutonomicon/cardano-transaction-lib/pull/1145/files#diff-bca190054ba134e1b1cd4ac4fda3466cddc5091bbc6798d2a08c8b20d70d6f66R136

Not sure what this it the best way to access `PlutusScript.bytes`, but I found no other.

File: src/Internal/Serialization/Types.purs [hunk 1]
```diff
@@ -132,7 +133,9 @@ foreign import data PlutusData :: Type
 foreign import data PlutusList :: Type
 foreign import data PlutusMap :: Type
 foreign import data PlutusScript :: Type
+foreign import bytes :: PlutusScript -> ByteArray
```
Comment (klntsky) [inline src/Internal/Serialization/Types.purs:136]: This function shouldn't be here.

File: src/Internal/Cardano/TextEnvelope.purs [hunk 2]
```diff
@@ -1,15 +1,12 @@
 module Ctl.Internal.Cardano.TextEnvelope
-  ( decodeTextEnvelope
+  ( TextEnvelope(..)
```
Comment (klntsky) [inline src/Internal/Cardano/TextEnvelope.purs:2]: Add `make check-format` to pre-commit hooks

File: src/Internal/Cardano/TextEnvelope.purs [hunk 3]
```diff
@@ -74,23 +82,29 @@ derive instance Newtype TextEnvelope _
 data TextEnvelopeDecodeError
   = JsonDecodeError JsonDecodeError
   | CborParseError CborParseError
+  | CSLError String -- Error produced by Cardano-serialization-lib
+
+decodeCborHexToBytes :: String -> Either TextEnvelopeDecodeError ByteArray
+decodeCborHexToBytes cborHex = do
+  cborBa <- note (JsonDecodeError $ TypeMismatch "Hex") $ hexToByteArray cborHex
+  -- NOTE PlutusScriptV1 is doubly-encoded cbor, so the resulting `ByteArray`
```
Comment (klntsky) [inline src/Internal/Cardano/TextEnvelope.purs:90]: No need for this comment here

File: src/Internal/Cardano/TextEnvelope.purs [hunk 4]
```diff
@@ -74,23 +82,29 @@ derive instance Newtype TextEnvelope _
...
-  -- https://github.com/Emurgo/cardano-serialization-lib/issues/268#issuecomment-1042986055
-  ba <- lmap CborParseError $ toByteArray $ wrap $ wrap cborBa
+  ba <- decodeCborHexToBytes cborHex
   pure $ wrap { type_, description, bytes: ba }
 
 printTextEnvelopeDecodeError :: TextEnvelopeDecodeError -> String
 printTextEnvelopeDecodeError = case _ of
   JsonDecodeError err -> printJsonDecodeError err
   CborParseError err -> show err
+  CSLError errMessage -> errMessage
```
Comment (klntsky) [inline src/Internal/Cardano/TextEnvelope.purs:107]: It should be `CslError`

File: src/Internal/Cardano/TextEnvelope.purs [hunk 5]
```diff
@@ -100,3 +114,25 @@ textEnvelopeBytes json ty = do
...
+  -> Either TextEnvelopeDecodeError PlutusScript
+_plutusScriptVxFromEnvelope version bytesToScript str = do
+  bytes <- textEnvelopeBytes str version
+  bimap
+    (CSLError <<< match { fromBytesError: \x -> x })
+    (bytesToScript <<< ST.bytes)
+    (fromBytes' bytes :: E (FromBytesError + ()) ST.PlutusScript)
+
+plutusScriptV1FromEnvelope
+  :: String -> Either TextEnvelopeDecodeError PlutusScript
```
Comment (klntsky) [inline src/Internal/Cardano/TextEnvelope.purs:131]: This interface is overcomplicated. I don't think the average user cares about the reason envelope parsing has failed and wants to pattern match on the reason. Please change it to just `Maybe` here and everywhere.

Comment (klntsky) [reply src/Internal/Cardano/TextEnvelope.purs:131]: Remove the lifting function as well, because we already have lifting for Maybe

**Review by klntsky — CHANGES_REQUESTED** [review]

File: src/Internal/Cardano/TextEnvelope.purs (hunk 5 above)
Comment (klntsky) [reply src/Internal/Cardano/TextEnvelope.purs:131]: And the error type itself

File: src/Internal/Serialization/Types.purs (hunk 1 above)
Comment (uhbif19) [reply src/Internal/Serialization/Types.purs:136]: Where does it should be?

File: src/Internal/Serialization/PlutusScript.purs [hunk 6]
```diff
@@ -12,6 +13,8 @@ import Ctl.Internal.Types.Scripts
   ) as T
 import Data.Tuple.Nested ((/\))
 
+foreign import bytes :: PlutusScript -> ByteArray
```
Comment (klntsky) [inline src/Internal/Serialization/PlutusScript.purs:16]: The name should be changed to plutusScriptBytes.

**Review by klntsky — CHANGES_REQUESTED** [review]
All FFI code should be hot in the tests. Please add a test - just a fixture would be fine.

**Comment by uhbif19** [comment]
@klntsky 

> All FFI code should be hot in the tests. Please add a test - just a fixture would be fine.

Is not it already covered by running examples in Plutip tests?

File: src/Internal/Cardano/TextEnvelope.purs [hunk 7]
```diff
@@ -71,32 +80,43 @@ newtype TextEnvelope =
...
 textEnvelopeBytes json ty = do
-  aeson <- lmap JsonDecodeError $ parseJsonStringToAeson json
+  aeson <- hush $ parseJsonStringToAeson json
   TextEnvelope te <- decodeTextEnvelope aeson
-  unless (te.type_ == ty) $ throwError $ JsonDecodeError $ TypeMismatch $
-    show ty
-  pure te.bytes
+  if (te.type_ == ty) then pure te.bytes else Nothing
+
+_plutusScriptVxFromEnvelope
```
Comment (klntsky) [inline src/Internal/Cardano/TextEnvelope.purs:103]:
```suggestion
plutusScriptFromEnvelope
```

File: src/Internal/Cardano/TextEnvelope.purs [hunk 8]
```diff
@@ -71,32 +80,43 @@ newtype TextEnvelope =
...
+  map
+    (bytesToScript <<< plutusScriptBytes)
+    (hush ((fromBytes' bytes) :: E (FromBytesError + ()) ST.PlutusScript))
+
+plutusScriptV1FromEnvelope
+  :: String -> Maybe PlutusScript
+plutusScriptV1FromEnvelope str = do
+  _plutusScriptVxFromEnvelope PlutusScriptV1 plutusV1Script str
+
+plutusScriptV2FromEnvelope
```
Comment (klntsky) [inline src/Internal/Cardano/TextEnvelope.purs:119]: Let's make these functions accept a `TextEnvelope`. That way we will check the validity of the envelope and make it clear how to use this API (the user will have to just follow the types).

File: src/Internal/Cardano/TextEnvelope.purs [hunk 9]
```diff
@@ -71,32 +80,43 @@ newtype TextEnvelope =
...
-    show ty
-  pure te.bytes
+  if (te.type_ == ty) then pure te.bytes else Nothing
+
+_plutusScriptVxFromEnvelope
+  :: TextEnvelopeType
+  -> (ByteArray -> PlutusScript)
+  -> String
+  -> Maybe PlutusScript
+_plutusScriptVxFromEnvelope version bytesToScript str = do
```
Comment (klntsky) [inline src/Internal/Cardano/TextEnvelope.purs:108]: Within our conventions/code style we use underscores only for FFI functions and lenses.

**Review by klntsky — CHANGES_REQUESTED** [review]
I think we should either remove the TextEnvelopeType tag (replacing it with a `String`), or extend it with `Other String` to allow using `textEnvelopeBytes` even with types we don't have separate functions for, to at least be able to get their byte representations from the envelope.

File: src/Internal/Cardano/TextEnvelope.purs (hunk 8 above)
Comment (uhbif19) [reply src/Internal/Cardano/TextEnvelope.purs:119]: But this way examples will be full of boilerplate.

**Comment by uhbif19** [comment]
`FromBytes` for `PlutusScript` somehow breaks code, so I converted bytes without it like it was in the code before.

File: src/Contract/TextEnvelope.purs [hunk 10]
```diff
@@ -1,30 +1,15 @@
...
-import Data.Bifunctor (lmap)
-import Effect.Exception (error)
-
-textEnvelopeBytes :: String -> TE.TextEnvelopeType -> Contract () ByteArray
-textEnvelopeBytes json ty =
-  liftEither $ lmap (error <<< printTextEnvelopeDecodeError) $
-    TE.textEnvelopeBytes json ty
+  , plutusScriptV1FromEnvelope
+  , plutusScriptV2FromEnvelope
+  ) as X
```
Comment (klntsky) [inline src/Contract/TextEnvelope.purs:15]: `PaymentSigningKeyShelleyed25519` and `StakeSigningKeyShelleyed25519` should have their corresponding functions too.

**Review by klntsky — CHANGES_REQUESTED** [review]

File: src/Contract/TextEnvelope.purs (hunk 10 above)
Comment (uhbif19) [reply src/Contract/TextEnvelope.purs:15]: There are such functions in `Ctl.Internal.Wallet.KeyFile` but they work with strings. I think I should make them work with `TextEnvelope`? Maybe reexport here in `Contract.TextEnvelope`?

Comment (uhbif19) [reply src/Contract/TextEnvelope.purs:15]: Oh, I seen old branch, I already fixed the types. So the only question is if to reexport them.

Comment (uhbif19) [reply src/Contract/TextEnvelope.purs:15]: @klntsky What I should do about it?

Comment (klntsky) [reply src/Contract/TextEnvelope.purs:15]: Thanks, if they accept the envelope type and not string, it's all good.

**Review by klntsky — APPROVED** [review]

File: CHANGELOG.md [hunk 11]
```diff
@@ -63,6 +63,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - `MintingPolicy` to an enum consisting of `PlutusScript` or `NativeScript` ([#1069](https://github.com/Plutonomicon/cardano-transaction-lib/pull/1069))
 - `Contract.Scripts` `applyArgs` is now monomorphic on the script parameter ([#1069](https://github.com/Plutonomicon/cardano-transaction-lib/pull/1069))
 - Adapted Gero wallet extension to `preview` network in E2E test suite ([#1086](https://github.com/Plutonomicon/cardano-transaction-lib/pull/1086))
+- `Contact.TextEnvelope` how exports more type safe `plutusScriptV1FromEnvelope` instread of `textEnvelopeBytes`, and uses `Maybe` instead of `Either` for error handling ([#988](https://github.com/Plutonomicon/cardano-transaction-lib/issues/988))
```
Comment (klntsky) [inline CHANGELOG.md:66]:
```suggestion
- `Contact.TextEnvelope` how provides more type safe interface with simplified error handling ([#988](https://github.com/Plutonomicon/cardano-transaction-lib/issues/988))
```

**Pull request merged**